### PR TITLE
Liouvillian cases for Parametric Risch differential equation

### PR DESCRIPTION
**Pull request by gxyd**

### CODE REVIEW

File: sympy/integrals/prde.py [hunk 1]
```diff
@@ -441,6 +441,42 @@ def prde_no_cancel_b_small(b, Q, n, DE):
...
+    log(x + 1)*log(x), log(x + 1)/log(x)
+
+    """
+    m = len(Q)
+    f, A = [None]*(n + 1), [None]*(n + 1)
+
+    # Why use DecrementLevel? Below line answers that:
+    # Assuming that we can solve such problems over 'k' (not k[t])
+    with DecrementLevel(DE):
+        DE0 = DE
```
Comment (asmeurer) [inline sympy/integrals/prde.py]: This doesn't seem proper use of the context manager. Is DE0 actually assigned properly below?

Comment (gxyd) [inline sympy/integrals/prde.py]: > This doesn't seem proper use of the context manager.

No it isn't. Kalevi told me about it (the mutability issue).

File: sympy/integrals/prde.py [hunk 2]
```diff
@@ -471,23 +507,28 @@ def param_poly_rischDE(a, b, q, n, DE):
...
-        elif ((b.is_zero or b.degree() < DE.d.degree() - 1)
-                and (DE.case == 'base' or DE.d.degree() >= 2)):
+        elif ((b.is_zero or b.degree(DE.t) < DE.d.degree(DE.t) - 1)
+                and (DE.case == 'base' or DE.d.degree(DE.t) >= 2)):
             return prde_no_cancel_b_small(b, q, n, DE)
 
-        elif (DE.d.degree() >= 2 and
-              b.degree() == DE.d.degree() - 1 and
+        elif (DE.d.degree(DE.t) >= 2 and
+              b.degree(DE.t) == DE.d.degree(DE.t) - 1 and
```
Comment (asmeurer) [inline sympy/integrals/prde.py]: Why these changes? Every polynomial should already be in that variable.

File: sympy/integrals/prde.py [hunk 3]
```diff
@@ -471,23 +507,28 @@ def param_poly_rischDE(a, b, q, n, DE):
...
               n > -b.as_poly(DE.t).LC()/DE.d.as_poly(DE.t).LC()):
             raise NotImplementedError("prde_no_cancel_b_equal() is "
                 "not yet implemented.")
 
         else:
-            # Cancellation
-            raise NotImplementedError("prde_cancel() is not yet "
-                "implemented.")
+            if b.is_zero:
+                pass
```
Comment (asmeurer) [inline sympy/integrals/prde.py]: Should this be NotImplementedError?

File: sympy/integrals/prde.py (hunk 1 above)
Comment (gxyd) [inline sympy/integrals/prde.py]: Also, do you mean that programmatically it is wrong to ever make a "deep copy" via context manager? (I was trying to do that)

Comment (asmeurer) [inline sympy/integrals/prde.py]: Is there not a method on DE to get a DE one level lower? The context manager is supposed to be for when a group of calculations are all only on the lower level (usually recursively).

Comment (asmeurer) [inline sympy/integrals/prde.py]: Although we perhaps should change it to not mutate and return a new DE in the context manager anyway.

Comment (asmeurer) [inline sympy/integrals/prde.py]: OK, I took a look at the code (I was on a plane when I wrote the above so I couldn't check). Perhaps the API should be modified so that DifferentialExtension is immutable. The API of risch is all internal, so there are no issues with changing it. In fact, I would highly expect it to need to change as progress is made in the algorithm.

So I suggest:

- Modify `decrement_level` and `increment_level` to return a new DifferentialExtension object. It should do so without redoing any unnecessary calculations. 
- `with DecrementLevel(DE)` would be changed to `with DecrementLevel(DE) as DE0`.  This is perhaps more annoying, as you have to use `DE0` inside the context manager instead of `DE`, but it would let you mix the original DE with the decremented one (which you apparently want to do here).

File: sympy/integrals/prde.py [hunk 4]
```diff
@@ -441,6 +441,49 @@ def prde_no_cancel_b_small(b, Q, n, DE):
...
+    #H = [[Poly(0, DE.t)]*m for i in range(n + 1)]
+
+    # Why use DecrementLevel? Below line answers that:
+    # Assuming that we can solve such problems over 'k' (not k[t])
+    if DE.case == 'primitive':
+        with DecrementLevel(DE):
+            ba, bd = frac_in(b, DE.t, field=True)
+    elif DE.case == 'exp':
+        with DecrementLevel(DE):
+            ba, bd = frac_in(b + n*derivation(DE.t, DE)/DE.t, DE.t)
```
Comment (jksuom) [inline sympy/integrals/prde.py]: We should probably have `field=True` also here.

File: sympy/integrals/prde.py [hunk 5]
```diff
@@ -459,24 +459,25 @@ def prde_cancel_liouvillian(b, Q, n, DE):
...
             Qyn = [frac_in(q.nth(N), DE.t) for q in Q]
-            f[N], A[N] = param_rischDE(ba, bd, Qyn, DE)
-        f[N] = [Poly(fa.as_expr()/fd.as_expr(), DE.t, field=True)
-                for fa, fd in f[N]]
-
-    r = [v.nullspace() for v in A]
-    # take the first vector of nullspace
-    d = [i[0][m - 1:] for i in r]
-
-    for N in range(n, -1, -1):
```
Comment (gxyd) [inline sympy/integrals/prde.py]: Can the two loops be put together as is I did here?

File: sympy/integrals/prde.py [hunk 6]
```diff
@@ -441,6 +441,49 @@ def prde_no_cancel_b_small(b, Q, n, DE):
...
+    if DE.case == 'primitive':
+        with DecrementLevel(DE):
+            ba, bd = frac_in(b, DE.t, field=True)
+    elif DE.case == 'exp':
+        with DecrementLevel(DE):
+            ba, bd = frac_in(b + n*derivation(DE.t, DE)/DE.t, DE.t)
+
+    for N in range(n, -1, -1):
+        with DecrementLevel(DE):
+            Qyn = [frac_in(q.nth(N), DE.t) for q in Q]
```
Comment (jksuom) [inline sympy/integrals/prde.py]: Note that this list shall be extended after each pass as in the equation on top of p. 238.

File: sympy/integrals/prde.py (hunk 5 above)
Comment (jksuom) [inline sympy/integrals/prde.py]: I cannot find this in the Files changed list.

Comment (gxyd) [inline sympy/integrals/prde.py]: This may not be there in overall files changed list. But seeing this particular commit or other way would be seeing [2nd commit](https://github.com/sympy/sympy/pull/12743/commits/0c8f943d91aada4ac5432056396cbd6d7b99cd12) side-by-side with [3rd commit](https://github.com/sympy/sympy/pull/12743/commits/08a7adc2c266429e3ccc5e5ac3b4d2a5dc3d39a3) of this pull request.

File: sympy/integrals/prde.py [hunk 7]
```diff
@@ -441,6 +441,50 @@ def prde_no_cancel_b_small(b, Q, n, DE):
...
+            fn, An = param_rischDE(ba, bd, Qyn, DE)
+        fn = [Poly(fa.as_expr()/fd.as_expr(), DE.t, field=True)
+                for fa, fd in fn]
+
+        V = An.nullspace()
+        # take the first vector of nullspace
+        # v = [c1, ..., cm, d1n, ..., drn_n]
+        v = An[0]
```
Comment (jksuom) [inline sympy/integrals/prde.py]: It is too early to pick a solution. All options should be left available until the solutions of param_rischDE have been found for all powers of t: `t**n, ..., t**i, ..., t**0`. Then the results are combined to a big matrix. which is returned together with a list of solution components `h_ji = f_ji t**i`.

File: sympy/integrals/prde.py (hunk 1 above)
Comment (gxyd) [inline sympy/integrals/prde.py]: I was trying to do what you suggested, but I don't think it would be wise to make such a change, comment by Kalevi very much sums that up I think https://gitter.im/jksuom/sympy?at=5940c478142826e972b28a06 ,

- DifferentialExtension object involves `Poly`, `list`, `GeneratorType`, etc,..., the first three of them require different methods to create a deep-copy of them, `copy.deepcopy(list)` works, but for `GeneratorType`, `Poly` it wouldn't work. We can create a deep-copy but the code would look messy and hackish.

> it would let you mix the original DE with the decremented one (which you apparently want to do here).

We can do without that for now.

Comment (asmeurer) [inline sympy/integrals/prde.py]: You don't need to deep copy Poly, as it's an immutable type. The only thing that would need to be explicitly copied are mutable data structures such as lists, and even then, only if they ever get mutated (probably we could change them to tuples so that it is clear that they aren't).

File: sympy/integrals/prde.py [hunk 8]
```diff
@@ -441,6 +441,59 @@ def prde_no_cancel_b_small(b, Q, n, DE):
...
+        with DecrementLevel(DE):
+            ba, bd = frac_in(b + n*derivation(DE.t, DE)/DE.t, DE.t, field=True)
+
+    Qy = [None]*(n + 1)
+    for N in range(n, -1, -1):
+        with DecrementLevel(DE):
+            Qyn = [frac_in(q.nth(N), DE.t, field=True) for q in Q]
+            Qy[N] = Qyn
+            fn, An = param_rischDE(ba, bd, Qyn, DE)
+        f[N] = [Poly(fa.as_expr()/fd.as_expr(), DE.t, field=True)
```
Comment (jksuom) [inline sympy/integrals/prde.py]: It seems that the elements of `fn` will be needed in two forms, as input pairs `fa, fd` for the following calls of `param_rischDE` and as expressions in the final output list of this routine.

File: sympy/integrals/prde.py (hunk 6 above)
Comment (gxyd) [inline sympy/integrals/prde.py]: Is my latest commit in accordance with your comment?

File: sympy/integrals/prde.py [hunk 9]
```diff
@@ -441,6 +441,52 @@ def prde_no_cancel_b_small(b, Q, n, DE):
...
+
+        ri = len(fn)
+        Fi = F[i] = [None]*ri
+
+        for j in range(ri):
+            hji = f[i][j]*DE.t**i
+            Fi[j] = derivation(hji, DE) - b*hji
+        # in the next loop instead of Qy it has
+        # to be Qy + Fi taking its place, like below?
+        Qy += Fi
```
Comment (gxyd) [inline sympy/integrals/prde.py]: Should this be `Q = Q + Fi`, instead?

Comment (jksuom) [inline sympy/integrals/prde.py]: I think the difference is that `Q + Fi` creates a new list while `Q += Fi` would extend the original. I don't know if that matters.

Comment (gxyd) [inline sympy/integrals/prde.py]: The point is: should it be `Q` instead of `Qy`? I am sorry, I didn't ask it the right way.

Comment (jksuom) [inline sympy/integrals/prde.py]: If `Q` is the original list and `Qy` is the extended list then this should be `Qy`. But it seems that the extended list is also denoted by `Qyi` above. Are they supposed to be different?

Comment (gxyd) [inline sympy/integrals/prde.py]: I have removed any usage of `Qy` or `Qyi` and extended the original `Q` itself. Do things look okay now?

**Comment by gxyd** [comment]
We have the integral `risch_integrate(log(log(x + exp(x))), x)` raising the error:
```
In [2]: risch_integrate(log(log(x + exp(x))), x)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-0845f081550f> in <module>()
----> 1 risch_integrate(log(log(x + exp(x))), x)

/home/gxyd/foss/sympy/sympy/integrals/risch.py in risch_integrate(f, x, extension, handle_first, separate_integral, rewrite_complex, conds)
   1708             ans, i, b = integrate_hyperexponential(fa, fd, DE, conds=conds)
   1709         elif case == 'primitive':
-> 1710             ans, i, b = integrate_primitive(fa, fd, DE)
   1711         elif case == 'base':
   1712             # XXX: We can't call ratint() directly here because it doesn't

/home/gxyd/foss/sympy/sympy/integrals/risch.py in integrate_primitive(a, d, DE, z)
   1355     p = p.as_poly(DE.t)
   1356 
-> 1357     q, i, b = integrate_primitive_polynomial(p, DE)
   1358 
   1359     ret = ((g1[0].as_expr()/g1[1].as_expr() + q.as_expr()).subs(s) +

/home/gxyd/foss/sympy/sympy/integrals/risch.py in integrate_primitive_polynomial(p, DE)
   1307 
   1308             try:
-> 1309                 rv = limited_integrate(aa, ad, [(Dta, Dtb)], DE)
   1310                 if rv is None:
   1311                     raise NonElementaryIntegralException

/home/gxyd/foss/sympy/sympy/integrals/prde.py in limited_integrate(fa, fd, G, DE)
    807     Fd = Poly(1, DE.t)
    808     G = [(fa, fd)] + G
--> 809     h, A = param_rischDE(Fa, Fd, G, DE)
    810     V = A.nullspace()
    811     V = [v for v in V if v[0] != 0]

/home/gxyd/foss/sympy/sympy/integrals/prde.py in param_rischDE(fa, fd, G, DE)
    707         n = oo
    708 
--> 709     h, B = param_poly_rischDE(a, b, r, n, DE)
    710 
    711     # h = [h1, ..., hv] in k[t]^v and and B is a matrix with u + v

/home/gxyd/foss/sympy/sympy/integrals/prde.py in param_poly_rischDE(a, b, q, n, DE)
    535             # Liouvillian cases
    536             if DE.case == 'primitive' or DE.case == 'exp':
--> 537                 return prde_cancel_liouvillian(b, q, n, DE)
    538             else:
    539                 raise NotImplementedError("non-linear and hypertangent "

/home/gxyd/foss/sympy/sympy/integrals/prde.py in prde_cancel_liouvillian(b, Q, n, DE)
    454             ba, bd = frac_in(b, DE.t, field=True)
    455 
--> 456     for i in range(n, -1, -1):
    457         if DE.case == 'exp': # this re-checking can be avoided
    458             with DecrementLevel(DE):

TypeError: 'Infinity' object cannot be interpreted as an integer
```
I remember you mentioning about it. For wolfram, it "exceeds standard computation time".

**Comment by jksuom** [comment]
Apparently [this line](https://github.com/sympy/sympy/blob/master/sympy/integrals/prde.py#L657) should be changed to something finite. It should also be examined if `bound_degree` could be improved.

**Comment by gxyd** [comment]
> Apparently this line should be changed to something finite. It should also be examined if bound_degree could be improved.

Should we address that in this PR?

**Comment by jksuom** [comment]
I think that n should be made finite (and not too big) in this PR to enable testing.

**Comment by gxyd** [comment]
It seems that it can't be tested with the above test case, for any finite value of `n`:
```
In [2]: t0, t1 = symbols('t0, t1')

In [3]: fa, fd = Poly(0, t1, domain='ZZ'), Poly(1, t1, domain='ZZ')

In [4]: G = [(Poly(1, t1, domain='ZZ'), Poly(1, t1, domain='QQ')), (Poly(t0 + 1, t1, domain='ZZ[t0]'), Poly((t0 + x)*t1, t1, domain='ZZ[x,t0]')
   ...: )]

In [5]: DE = DifferentialExtension(extension={'D': [Poly(1, x, domain='ZZ'), Poly(t0, t0, domain='ZZ'), Poly((t0 + 1)/(t0 + x), t1, domain='ZZ(
   ...: x, t0)')]})

In [6]: from sympy.integrals.prde import param_rischDE

In [7]: param_rischDE(fa, fd, G, DE)
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-7-3e0ac41e6a34> in <module>()
----> 1 param_rischDE(fa, fd, G, DE)

/home/gxyd/foss/sympy/sympy/integrals/prde.py in param_rischDE(fa, fd, G, DE)
    707         n = 0
    708 
--> 709     h, B = param_poly_rischDE(a, b, r, n, DE)
    710 
    711     # h = [h1, ..., hv] in k[t]^v and and B is a matrix with u + v

/home/gxyd/foss/sympy/sympy/integrals/prde.py in param_poly_rischDE(a, b, q, n, DE)
    535             # Liouvillian cases
    536             if DE.case == 'primitive' or DE.case == 'exp':
--> 537                 return prde_cancel_liouvillian(b, q, n, DE)
    538             else:
    539                 raise NotImplementedError("non-linear and hypertangent "

/home/gxyd/foss/sympy/sympy/integrals/prde.py in prde_cancel_liouvillian(b, Q, n, DE)
    461         with DecrementLevel(DE):
    462             Qy = [frac_in(q.nth(i), DE.t, field=True) for q in Q]
--> 463             fi, Ai = param_rischDE(ba, bd, Qy, DE)
    464         fi = [Poly(fa.as_expr()/fd.as_expr(), DE.t, field=True)
    465                 for fa, fd in fi]

/home/gxyd/foss/sympy/sympy/integrals/prde.py in param_rischDE(fa, fd, G, DE)
    649     gamma *= hn
    650 
--> 651     A, B, G, hs = prde_special_denom(a, ba, bd, G, DE)
    652     # Solutions p in k[t] of  A*Dp + B*p = Sum(ci*Gi) correspond
    653     # to solutions q = p/hs of the previous equation.

/home/gxyd/foss/sympy/sympy/integrals/prde.py in prde_special_denom(a, ba, bd, G, DE, case)
    131                 alphaa, alphad = frac_in(-ba.eval(0)/bd.eval(0)/a.eval(0), DE.t)
    132                 etaa, etad = frac_in(dcoeff, DE.t)
--> 133                 A = parametric_log_deriv(alphaa, alphad, etaa, etad, DE)
    134                 if A is not None:
    135                     a, m, z = A

/home/gxyd/foss/sympy/sympy/integrals/prde.py in parametric_log_deriv(fa, fd, wa, wd, DE)
    920     # TODO: Write the full algorithm using the structure theorems.
    921 #    try:
--> 922     A = parametric_log_deriv_heu(fa, fd, wa, wd, DE)
    923 #    except NotImplementedError:
    924         # Heuristic failed, we have to use the full method.

/home/gxyd/foss/sympy/sympy/integrals/prde.py in parametric_log_deriv_heu(fa, fd, wa, wd, DE, c1)
    888 
    889     if not z.has(DE.t):
--> 890         raise NotImplementedError("parametric_log_deriv_heu() "
    891             "heuristic failed: z in k.")
    892 

NotImplementedError: parametric_log_deriv_heu() heuristic failed: z in k.
```

But I have added the test case involved in the calculation of `risch_integrate(log(log(1 + exp(x))), x)`, which is correctly evaluated after setting `n = 2`.

File: sympy/integrals/prde.py [hunk 10]
```diff
@@ -652,9 +702,11 @@ def param_rischDE(fa, fd, G, DE):
         # it will always terminate no matter what n is.
         n = bound_degree(a, b, r, DE, parametric=True)
     except NotImplementedError:
-        debug("param_rischDE: Proceeding with n = oo; may cause "
```
Comment (gxyd) [inline sympy/integrals/prde.py]: I think we can remove this debug line, if we set some finite value, will do in next commit.

Comment (jksuom) [inline sympy/integrals/prde.py]: Perhaps it should be replaced with a suitable comment. A temporary bound is set. Eventually, it will be removed.

**Comment by jksuom** [comment]
> NotImplementedError: parametric_log_deriv_heu() heuristic failed: z in k.

We should probably treat this as 'no solution' for the present until the structure theorem version of `parametric_log_deriv` is implemented. That should suffice for testing.

File: sympy/integrals/tests/test_prde.py [hunk 11]
```diff
@@ -270,13 +327,3 @@ def test_parametric_log_deriv():
     assert parametric_log_deriv_heu(Poly(5*t**2 + t - 6, t), Poly(2*x*t**2, t),
     Poly(-1, t), Poly(x*t**2, t), DE) == \
         (2, 6, t*x**5)
-
-
-def test_issue_10798():
```
Comment (asmeurer) [inline sympy/integrals/tests/test_prde.py]: Why was this test removed? It seems it is in the wrong file.

Comment (gxyd) [inline sympy/integrals/tests/test_prde.py]: The test has **not** been removed, it has been shifted to [here](https://github.com/sympy/sympy/pull/12743/files#diff-af61f7566bec5996f0ce919794ab7413R152), i.e. in the test function `test_prde_cancel_liouvillian`.

The test is present is in right file in the sense that it is testing issue #10798 which was an issue with `is_deriv_k` though for that issue a more lower-level test could have been added i.e. we can replace the above test of `risch_integrate` with:
```
>>> fa, fd = Poly(-1, t0, domain='ZZ'), Poly(y, t0, domain='ZZ[y]')
>>> DE = DifferentialExtension(extension={'D': [Poly(1, y, domain='ZZ'), Poly(-1/y, t0, domain='ZZ(y)')], 'E_K': [], 'E_args': [], 'L_K': [1], 'L_args': [1/y]})
>>> is_deriv_k(fa, fd, DE)
([(t0, 1)], t0, -1)
```

But my point to keep the original test of `risch_integrate` is: as it tests the issue #10798 also it tests the `prde_cancel()`.

Comment (gxyd) [inline sympy/integrals/tests/test_prde.py]: As I could see the test I mentioned was already added by Kalevi (in a later PR for that issue). And here is this modified test in the right file (i.e `test_risch.py`) https://github.com/sympy/sympy/pull/12743/commits/616b0a0db0adcb9f3201b89c5f25cd6c933a187e#diff-72dfd0e04e99c9981ab3d8349b5eaa09R646

File: sympy/integrals/tests/test_prde.py [hunk 12]
```diff
@@ -144,6 +146,41 @@ def test_prde_no_cancel():
     assert (Matrix([q])*V[0][:6, :])[0] == Poly(x - 1/2, t, domain='QQ(x)')
 
 
+def test_prde_cancel_liouvillian():
+    ### 1. case == 'primitive'
+    # used when integrating (log(-1/y)/2 - log(1/y)/2)/y - (log(1 - 1/y)/2 - log(1 + 1/y)/2)/y
+    # Not taken from 'the' book
+    DE = DifferentialExtension(extension={'D': [Poly(1, x), Poly(1/x, t)]})
+    assert prde_cancel_liouvillian(Poly(-1/(x - 1), t), [Poly(-x + 1, t), Poly(1, t)], 1, DE) == \
```
Comment (jksuom) [inline sympy/integrals/tests/test_prde.py]: Does this turn up in the integration of an expression containing log(x)?

File: sympy/integrals/tests/test_prde.py [hunk 13]
```diff
@@ -144,6 +146,41 @@ def test_prde_no_cancel():
...
+        ([Poly(0, t, domain='QQ'), Poly(0, t, domain='QQ'), Poly((x - 1)*t, t, domain='ZZ(x)'),
+            Poly(0, t, domain='QQ'), Poly(0, t, domain='QQ'), Poly(0, t, domain='QQ'), Poly(0, t, domain='QQ'),
+            Poly(0, t, domain='QQ'), Poly(0, t, domain='QQ'), Poly(0, t, domain='QQ'), Poly(x - 1, t, domain='ZZ(x)'),
+            Poly(-x**2 + x, t, domain='ZZ(x)'), Poly(0, t, domain='QQ'), Poly(0, t, domain='QQ'), Poly(0, t, domain='QQ'),
+            Poly(0, t, domain='QQ')], Matrix([[ 0,  1,  0,  0, 0,  0,  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 ],
```
Comment (jksuom) [inline sympy/integrals/tests/test_prde.py]: Maybe the list and the matrix could be stored separately. Then some tests concerning rank and/or nullspace could be done with the matrix The expression itself is not very illuminating.

File: sympy/integrals/tests/test_prde.py (hunk 12 above)
Comment (gxyd) [inline sympy/integrals/tests/test_prde.py]: Yes, (that is a wrong comment above, I wonder how I made the mistake, it took me sometime to come up with what this integral is):
```
[gxyd@stallman sympy{liouvillian}]$ bin/isympy
IPython console for SymPy 1.0.1.dev (Python 3.6.0-64-bit) (ground types: python)

These commands were executed:
>>> from __future__ import division
>>> from sympy import *
>>> x, y, z, t = symbols('x y z t')
>>> k, m, n = symbols('k m n', integer=True)
>>> f, g, h = symbols('f g h', cls=Function)
>>> init_printing()

Documentation can be found at http://docs.sympy.org/dev


In [1]: import pdb

In [2]: from sympy.integrals.risch import risch_integrate, DifferentialExtension

In [3]: pdb.run('risch_integrate(log(x) - log(x - 1), x)')
> <string>(1)<module>()
(Pdb) s
--Call--
> /home/gxyd/foss/sympy/sympy/core/basic.py(105)__hash__()
-> def __hash__(self):
(Pdb) b sympy/integrals/prde.py:448
Breakpoint 1 at /home/gxyd/foss/sympy/sympy/integrals/prde.py:448
(Pdb) c
> /home/gxyd/foss/sympy/sympy/integrals/prde.py(448)prde_cancel_liouvillian()
-> H = []
(Pdb) p b
Poly(-1/(x - 1), _t0, domain='ZZ(x)')
(Pdb) p Q
[Poly(-x + 1, _t0, domain='ZZ(x)'), Poly(1, _t0, domain='ZZ(x)')]
(Pdb) p n
1
(Pdb) p DE
DifferentialExtension(dict([('f', log(x) - log(x - 1)), ('x', x), ('T', [x, _t0, _t1]), ('D', [Poly(1, x, domain='ZZ'), Poly(1/x, _t0, domain='ZZ(x)'), Poly(1/(x - 1), _t1, domain='ZZ(x)')]), ('fa', Poly(-_t1 + _t0, _t1, domain='ZZ[_t0]')), ('fd', Poly(1, _t1, domain='ZZ')), ('Tfuncs', [Lambda(_i, log(_i)), Lambda(_i, log(_i - 1))]), ('backsubs', []), ('E_K', []), ('E_args', []), ('L_K', [1, 2]), ('L_args', [x, x - 1]), ('cases', ['base', 'primitive', 'primitive']), ('case', 'primitive'), ('t', _t0), ('d', Poly(1/x, _t0, domain='ZZ(x)')), ('newf', _t0 - _t1), ('level', -2), ('dummy', True)]))
(Pdb) p DE.D
[Poly(1, x, domain='ZZ'), Poly(1/x, _t0, domain='ZZ(x)'), Poly(1/(x - 1), _t1, domain='ZZ(x)')]
(Pdb) p DE.level
-2
(Pdb) 
```

Comment (jksuom) [inline sympy/integrals/tests/test_prde.py]: It looks OK. Apparently the highest level with `D` given by `Poly(1/(x - 1), _t1, domain='ZZ(x)')` is not needed in this application of `prde_cancel_liouvillian`.

File: sympy/integrals/tests/test_risch.py [hunk 14]
```diff
@@ -636,6 +637,19 @@ def test_risch_integrate():
 
     assert risch_integrate(0, x) == 0
 
+    # also tests prde_cancel()
+    e1 = log(x/exp(x) + 1)
+    ans1 = risch_integrate(e1, x)
+    assert ans1 == (x*log(x*exp(-x) + 1) + NonElementaryIntegral((x**2 - x)/(x + exp(x)), x))
+    assert cancel(diff(ans1, x) - e1) == 0
```
Comment (gxyd) [inline sympy/integrals/tests/test_risch.py]: I've used `cancel` instead of `simplify`, as `cancel` is more specific (also taking a little less time). Also the end of 2nd para of https://github.com/sympy/sympy/issues/5109#issue-28950562 .

**Comment by gxyd** [comment]
I don't intend to make any more changes. I think the tests should also pass.
